crates/file_finder/src/file_finder.rs

```diff
@@ -522,84 +556,100 @@ impl FileFinderDelegate {
     ) {
         if search_id >= self.latest_search_id {
             self.latest_search_id = search_id;
-            let extend_old_matches = self.latest_search_did_cancel
-                && Some(query.path_like.path_query())
-                    == self
-                        .latest_search_query
-                        .as_ref()
-                        .map(|query| query.path_like.path_query());
+            let query_changed = self.query_changed(&query);
+            let extend_old_matches = self.latest_search_did_cancel && !query_changed;
             self.matches.push_new_matches(
                 &self.history_items,
                 self.currently_opened_path.as_ref(),
                 Some(&query),
                 matches.into_iter(),
                 extend_old_matches,
             );
+
+            if query_changed {
+                self.selected_match = None;
+            }
+
             self.latest_search_query = Some(query);
             self.latest_search_did_cancel = did_cancel;
-            self.selected_index = self.calculate_selected_index();
+            self.update_selected_index(self.calculate_selected_index());
+
             cx.notify();
         }
     }
 
+    fn query_changed(&self, new_query: &PathLikeWithPosition<FileSearchQuery>) -> bool {
+        return self
+            .latest_search_query
+            .as_ref()
+            .map(|q| q.path_like.raw_query != new_query.path_like.raw_query)
+            .unwrap_or(true);
+    }
+
+    fn update_selected_index(&mut self, selected_index: usize) {
+        self.selected_index = selected_index;
+        self.selected_match = self.matches.get(self.selected_index).map(|m| m.clone());
```

Comment:
NIT

```suggestion
        self.selected_match = self.matches.get(self.selected_index).cloned();
```

Reply:
Thank you for the review, both points make sense to me.

It's definitely worth to add a test, thanks for the reference.

In regards the readability - what I'm thinking of is moving `selected_index` and `selected_match` to the dedicated struct like `SelectedEntry`, since what we're trying to do here is to keep particular entry in the list selected even if it's index changes due to some events. Does it make sense to you?

I also wonder, if there's any particular reason for passing `selected` flag to `file_finder::render_match` instead of comparing `ix` to the `selected_index` stored inside the `file_finder`?

Reply:
> to the dedicated struct

I think that alone is not convincing, as leaves the same "why do we need to store both if matches[selected_index] is an O(1) access" question when reading, and current names do not help.
Ad-hoc weird options: `selected_by_user`, `selection_index_in_picker`.

Maybe it's not two fields but an enum that we need (weird ad-hoc ideas again)?
```rust
enum SelectedItem {
    UserSelected { i: usize } ,
    ReorderedAfterBackgroundUpdate { user_selected: usize, reordered_to: usize },
    SomethingElse,
}
```

I feel like there's a case where things have not diverged at all and there we can avoid any binary searches, etc.; and the other case, even if diverged, that could be expressed more explicitly.
I could be wrong though.


> any particular reason for passing selected flag to file_finder::render_match instead of comparing ix to the selected_index stored inside the file_finder

We kind of do that but compare with the delegate data: https://github.com/zed-industries/zed/blob/b6e28702520b0e4477de450052a48b0b1c625715/crates/picker/src/picker.rs#L499

The main reason for split is the fact that `FileFinder` encapsulates its render state inside and does not expose it — so various "go up" and other things change indices without forcing the user to write all that code.
While the `*Delegate` is where the real logic is, and changing the index there might influence whatever generic purpose, maybe even selecting multiple items or jumping to something that's not in `FileFinder`'s selection area.

Apart from that, I am not aware of any good reason for this besides an attempt to have a relatively low-level API for any kinds of use cases.

Reply:
I get the point, need to think a bit more about how to express the intent better.

Hopefully I'll revert with some updates during the week, but realistically, I'll get to it during the weekend, please, bear with me.

Reply:
No rush at all, I am not the fastest to reply here either. 
We might eventually come and close the PR if it's ~1 month or so but even that means nothing in the context of this PR — it's just a few more things we need to think on, nothing fundamentally wrong.

Reply:
Pushed some updates here - added a couple of tests, applied NITs.

The answer to the question "why do we need to store both if matches[selected_index] is an O(1) access" is that we actually don't need to. Reverted back to storing just the selected index, seems to still work ok. The question I have here: is it true, that if the match stays in the matches list, it still has the same score as before the update? Otherwise, it's quite complicated to apply binary search to check if the match is still present.

Please, let me know if the code now makes more sense to you.

Reply:
> The question I have here: is it true, that if the match stays in the matches list, it still has the same score as before the update? 

Depends on the score.

The `history_score` is derived from the order 

https://github.com/zed-industries/zed/blob/d93891ba630345d7e28b324aa027e650f4e7832f/crates/file_finder/src/file_finder.rs#L259

as we push the newly opened file on top constantly.

The regular score is derived once per `update_matches` call, when search matches are returned back in the `spawn_search` call.